src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,152 @@
+    private int size;
+
+    private class Node {
+        private T element;
+        private Node next;
+        private Node prev;
+
+        public Node(T element) {
+            this.element = element;
+        }
+    }
+
     @Override
     public void add(T value) {
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
+        Node newNode = new Node(value);
+        if (index == size) {
+            addNodeWhenIndexEqualsSize(newNode);
+        } else {
+            Node findNode = findNode(index, "Can't add by index " + index);
+            insertNode(findNode, newNode);
+        }
+    }
+
+    private void addNodeWhenIndexEqualsSize(Node newNode) {
+        if (size == 0) {
+            tail = head = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
+    }
+
+    private void insertNode(Node findNode, Node newNode) {
+        newNode.next = findNode;
+        newNode.prev = findNode.prev;
+        if (head == findNode) {
+            head = newNode;
+        } else {
+            findNode.prev.next = newNode;
+        }
+        findNode.prev = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (T item : list) {
+                add(item);
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node findNode = findNode(index, "Can't get by index " + index);
+        return findNode.element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node findNode = findNode(index, "Can't set by index " + index);
+        T tmp = findNode.element;
+        findNode.element = value;
+        return tmp;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node findNode = findNode(index, "Can't remove by index " + index);
+        return unlink(findNode);
     }
 
     @Override
     public boolean remove(T object) {
+        Node currentNode = head;
+        while (currentNode != null) {
+            if (object == currentNode.element
+                    || object != null
+                    && object.equals(currentNode.element)) {
+                unlink(currentNode);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node findNode(int index, String messageForException) {
+        if (size == 0 || index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(messageForException);
+        }
+        return (index <= size >> 1) ? getNodeStartHead(index) : getNodeStartTail(index);
```

Comment:
I am guessing that this is our check index method, despite being called `findNode` 
Do you remember how we did this last time? can we do it the same way here?  Methods should be responsible only for 1 thing, especially not accepting any exception messages in parameters